storage/src/main/java/io/delta/storage/GCSLogStore.java

```diff
@@ -0,0 +1,189 @@
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.LocalFileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.FileAlreadyExistsException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.stream.Stream;
+
+/**
+ * :: Unstable ::
+ * <p>
+ * The [[LogStore]] implementation for GCS, which uses gcs-connector to
+ * provide the necessary atomic and durability guarantees:
+ *
+ * <ol>
+ *   <li>Atomic Visibility: Read/read-after-metadata-update/delete are strongly
+ * consistent for GCS.</li>
+ *
+ *   <li>Consistent Listing: GCS guarantees strong consistency for both object and
+ * bucket listing operations.
+ * https://cloud.google.com/storage/docs/consistency</li>
+ *
+ *   <li>Mutual Exclusion: Preconditions are used to handle race conditions.</li>
+ * </ol>
+ * Regarding file creation, this implementation:
+ * <ul>
+ *    <li>Opens a stream to write to GCS otherwise.</li>
+ *    <li>Throws [[FileAlreadyExistsException]] if file exists and overwrite is false.</li>
+ *    <li>Assumes file writing to be all-or-nothing, irrespective of overwrite option.</li>
+ * </ul>
+ *
+ * @note This class is not meant for direct access but for configuration based on storage system.
+ * See https://docs.delta.io/latest/delta-storage.html for details.
+ */
+@InterfaceStability.Unstable
+public class GCSLogStore extends HadoopFileSystemLogStore {
+
+    String preconditionFailedExceptionMessage = "412 Precondition Failed";
+
+    public GCSLogStore(Configuration hadoopConf) {
+        super(hadoopConf);
+    }
+
+    private static <T> T runInNewThread(
+            String threadName,
+            Boolean isDaemon,
+            Callable<T> body) throws Exception {
+        List<Exception> exceptionHolder = new ArrayList<>(1);
+        List<T> resultHolder = new ArrayList<>(1);
+        Thread thread = new Thread(threadName) {
+            @Override
+            public void run() {
+                try {
+                    resultHolder.add(body.call());
+                } catch (Exception ex) {
+                    exceptionHolder.add(ex);
+                }
+            }
+        };
+        thread.setDaemon(isDaemon);
+        thread.start();
+        thread.join();
+        if (!exceptionHolder.isEmpty()) {
+            Exception realException = exceptionHolder.get(0);
+            // Remove the part of the stack that shows method calls into this helper method
+            // This means drop everything from the top until the stack element
+            // ThreadUtils.runInNewThread(), and then drop that as well (hence the `drop(1)`).
+            Stream<StackTraceElement> baseStackTrace = Arrays.stream(
+                    Thread.currentThread().getStackTrace()).dropWhile(t ->
```

Comment:
this `dropWhile` logic still needs to be converted into proper java code, right?

Reply:
Changed this .